src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -334,50 +334,90 @@ case class TableRepartition(child: TableIR, n: Int, shuffle: Boolean) extends Ta
   }
 }
 
-case class TableJoin(left: TableIR, right: TableIR, joinType: String) extends TableIR {
+object TableJoin {
+  def apply(left: TableIR, right: TableIR, joinType: String): TableJoin =
+    TableJoin(left, right, joinType, left.typ.keyOrEmpty.length)
+}
+
+/**
+  * Suppose 'left' has key [l_1, ..., l_n] and 'right' has key [r_1, ..., r_m].
+  * Then [l_1, ..., l_j] and [r_1, ..., r_j] must have the same type, where
+  * j = 'joinKey'. TableJoin computes the join of 'left' and 'right' along this
+  * common prefix of their keys, returning a table with key
+  * [l_1, ..., l_j, l_{j+1}, ..., l_n, r_{j+1}, ..., r_m] (with possible
+  * renaming of right field names to avoid collision).
+  *
+  * WARNING: If 'left' has any duplicate (full) key [k_1, ..., k_n], and j < m,
+  * and 'right' has multiple rows with the corresponding join key
+  * [k_1, ..., k_j] but distinct full keys, then the resulting table will have
+  * out-of-order keys. To avoid this, ensure one of the following:
+  *   * j == m
+  *   * 'left' has distinct keys
+  *   * 'right' has distinct join keys (length j prefix), or at least no
+  *     distinct keys with the same join key.
+  */
+case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: Int)
+  extends TableIR {
+
+  require(joinKey >= 0)
   require(left.typ.keyType.zip(right.typ.keyType).exists { case (leftKey, rightKey) =>
-    leftKey isIsomorphicTo rightKey
+    (leftKey.size >= joinKey) && (rightKey.size >= joinKey) &&
+      (leftKey.truncate(joinKey) isIsomorphicTo rightKey.truncate(joinKey))
   })
 
   val children: IndexedSeq[BaseIR] = Array(left, right)
 
-  private val joinedFields = left.typ.keyType.get.fields ++
-    left.typ.valueType.fields ++
-    right.typ.valueType.fields
+  private val leftRVDType =
+    OrderedRVDType(left.typ.key.get.take(joinKey), left.typ.rowType)
+  private val rightRVDType =
+    OrderedRVDType(right.typ.key.get.take(joinKey), right.typ.rowType)
+
+  private val joinedFields =
+    leftRVDType.kType.fields ++
+      leftRVDType.valueType.fields ++
+      rightRVDType.valueType.fields
   private val preNames = joinedFields.map(_.name).toArray
   private val (finalColumnNames, remapped) = mangle(preNames)
 
+  private val joinedGlobalFields = left.typ.globalType.fields ++ right.typ.globalType.fields
```

Comment:
I don't think this is really necessary, but happy to see the change.

Reply:
I talked to Cotton about this. We both thought this was conceptually the right thing to do, though not pressing. It was so easy to add that I just did it now while it was on my mind.